ethereum/p2p/src/test/java/tech/pegasys/pantheon/ethereum/p2p/discovery/PeerDiscoveryTestHelper.java

```diff
@@ -13,27 +13,207 @@
   }
 
-  public static DiscoveryPeer[] generatePeers(final SECP256K1.KeyPair... keypairs) {
-    return Stream.of(keypairs)
-        .map(kp -> kp.getPublicKey().getEncodedBytes())
-        .map(bytes -> new DiscoveryPeer(bytes, new Endpoint("127.0.0.1", 1, OptionalInt.empty())))
-        .toArray(DiscoveryPeer[]::new);
+  public MockPeerDiscoveryAgent startDiscoveryAgent(final DiscoveryPeer... bootstrapPeers) {
+    AgentBuilder agentBuilder = agentBuilder().setBootstrapPeers(bootstrapPeers);
+
+    return startDiscoveryAgent(agentBuilder);
+  }
+
+  /**
+   * Start a single discovery agent with the provided bootstrap peers.
+   *
+   * @param bootstrapPeers the list of bootstrap peers
+   * @param blacklist the peer blacklist
+   * @return a list of discovery agents.
+   */
+  public MockPeerDiscoveryAgent startDiscoveryAgent(
+      final List<DiscoveryPeer> bootstrapPeers, final PeerBlacklist blacklist) {
+    AgentBuilder agentBuilder =
+        agentBuilder().setBootstrapPeers(bootstrapPeers).setBlacklist(blacklist);
+
+    return startDiscoveryAgent(agentBuilder);
+  }
+
+  public MockPeerDiscoveryAgent startDiscoveryAgent(final AgentBuilder agentBuilder) {
+    final MockPeerDiscoveryAgent agent = createDiscoveryAgent(agentBuilder);
+    agent.start();
+    return agent;
+  }
+
+  public MockPeerDiscoveryAgent createDiscoveryAgent(final List<DiscoveryPeer> bootstrapPeers) {
+    AgentBuilder agentBuilder = agentBuilder().setBootstrapPeers(bootstrapPeers);
+
+    return createDiscoveryAgent(agentBuilder);
+  }
+
+  public MockPeerDiscoveryAgent createDiscoveryAgent(final DiscoveryPeer... bootstrapPeers) {
+    AgentBuilder agentBuilder = agentBuilder().setBootstrapPeers(bootstrapPeers);
+
+    return createDiscoveryAgent(agentBuilder);
   }
 
-  public static DiscoveryPeer[] generateDiscoveryPeers(final SECP256K1.KeyPair... keypairs) {
-    return Stream.of(generatePeers(keypairs)).map(DiscoveryPeer::new).toArray(DiscoveryPeer[]::new);
+  public MockPeerDiscoveryAgent createDiscoveryAgent(final AgentBuilder agentBuilder) {
+    final MockPeerDiscoveryAgent agent = agentBuilder.build();
+    agents.put(agent.getId(), agent);
+    return agent;
+  }
+
+  public static class AgentBuilder {
+    private final Map<BytesValue, MockPeerDiscoveryAgent> agents;
+    private final AtomicInteger nextAvailablePort;
+
+    private PeerBlacklist blacklist = new PeerBlacklist();
+    private NodeWhitelistController whitelist =
+        new NodeWhitelistController(PermissioningConfiguration.createDefault());
+    private List<DiscoveryPeer> bootstrapPeers = Collections.emptyList();
+    private boolean active = true;
+
+    public AgentBuilder(
+        final Map<BytesValue, MockPeerDiscoveryAgent> agents,
+        final AtomicInteger nextAvailablePort) {
+      this.agents = agents;
+      this.nextAvailablePort = nextAvailablePort;
+    }
+
+    public AgentBuilder setBootstrapPeers(final List<DiscoveryPeer> peers) {
```

Comment:
When I restored idiomatic getters/setters I didn't do setters for Builders (stuff that returned `this`).  For consistency could you change it to just <foo>?